### PR TITLE
HPCC-20959 Unable to query file scope permissions

### PR DESCRIPTION
**Pull request by RussWhitehead**
If sessions are enabled, and LDAP is enabled in Dali, and there
are no public/private key files configured, then permission requests
to Dali fail authentication. This PR authenticates the user if there is
no signature and the password is a valid HPCC workunit token

Signed-off-by: Russ Whitehead <william.whitehead@lexisnexis.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [x] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by RussWhitehead** [comment]
@wangkx Please review

File: dali/base/dasess.cpp [hunk 1]
```diff
@@ -947,9 +947,11 @@ class CClientSessionManager: public CSessionManagerBase, implements IConnectionM
...
                 CDateTime now;
                 StringBuffer b64sig;
-                createDaliSignature(obj, udesc, now, b64sig);
-                mb.append(b64sig.str());
-                now.serialize(mb);
+                if (createDaliSignature(obj, udesc, now, b64sig))
+                {
+                    mb.append(b64sig.str());
+                    now.serialize(mb);
+                }
```
Comment (wangkx) [inline dali/base/dasess.cpp]: Can the code handles the missing signature when the mb is read?

File: dali/server/daldap.cpp [hunk 2]
```diff
@@ -196,7 +197,7 @@ class CDaliLdapConnection: implements IDaliLdapConnection, public CInterface
                 ERRLOG("LDAP: getPermissions(%s) scope=%s user=%s digital signature support not available",key?key:"NULL",obj?obj:"NULL",username.str());
         }
 
-        if (!isEmptyString(user->credentials().getPassword()))
+        if (!isEmptyString(user->credentials().getPassword()) && !isWorkunitDAToken(user->credentials().getPassword()))
```
Comment (wangkx) [inline dali/server/daldap.cpp]: You verify the password only if it is not empty and it is not a WUDAToken, which is correct. But, what if the password is empty and the authentication is required?

**Review by wangkx — COMMENTED** [review]
@RussWhitehead 2 questions

File: dali/base/dasess.cpp (hunk 1 above)
Comment (RussWhitehead) [reply dali/base/dasess.cpp]: Yes, in processMessage, the code reads only if there is more

                StringBuffer reqSignature;
                CDateTime reqUTCTimestamp;
                if (mb.remaining() > 0)
                {
                    mb.read(reqSignature);
                    reqUTCTimestamp.deserialize(mb);
                }

File: dali/server/daldap.cpp (hunk 2 above)
Comment (RussWhitehead) [reply dali/server/daldap.cpp]: We discussed this at length with RKC, Jake, Tony and got Flavio's approval that a password will no longer be required in this call.  But if one is provided, it should be checked

**Comment by wangkx** [comment]
@RussWhitehead ok for me.

**Comment by RussWhitehead** [comment]
@richardkchapman Please merge

**Pull request merged**